hivemind/p2p/p2p_daemon.py

```diff
@@ -0,0 +1,352 @@
+            natPortMap=nat_port_map, autonat=auto_nat,
+            b=bootstrap, **{**bpeers, **dht, **freachability, **kwargs})
+        self._assign_daemon_ports(host_port, daemon_listen_port)
+        for try_count in range(self.NUM_RETRIES):
+            try:
+                self._initialize(proc_args)
+                await self._identify_client(P2P.RETRY_DELAY * (2 ** try_count))
+            except Exception as e:
+                logger.debug(f"Failed to initialize p2p daemon: {e}")
+                self._kill_child()
+                if try_count == P2P.NUM_RETRIES - 1:
+                    raise
+                self._assign_daemon_ports()
+                continue
+            break
+        return self
+
+    @classmethod
+    async def replicate(cls, daemon_listen_port: int, host_port: int):
+        self = cls()
+        # There is no child under control
+        # Use external already running p2pd
+        self._child = None
+        self._alive = True
+        self._assign_daemon_ports(host_port, daemon_listen_port)
+        self._client_listen_port = find_open_port()
+        self._client = p2pclient.Client(
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._daemon_listen_port}'),
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._client_listen_port}'))
+        await self._identify_client(0)
+        return self
+
+    async def wait_for_at_least_n_peers(self, n_peers, attempts=3):
+        for _ in range(attempts):
+            peers = await self._client.list_peers()
+            if len(peers) >= n_peers:
+                return
+            await asyncio.sleep(1)
+
+        raise RuntimeError('Not enough peers')
+
+    def _initialize(self, proc_args: tp.List[str]) -> None:
+        proc_args = copy.deepcopy(proc_args)
+        proc_args.extend(self._make_process_args(
+            hostAddrs=f'/ip4/0.0.0.0/tcp/{self._host_port},/ip4/0.0.0.0/udp/{self._host_port}/quic',
+            listen=f'/ip4/127.0.0.1/tcp/{self._daemon_listen_port}'
+        ))
+        self._child = subprocess.Popen(
+            args=proc_args,
+            stdin=subprocess.PIPE, stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE, encoding="utf8"
+        )
+        self._alive = True
+        self._client_listen_port = find_open_port()
+        self._client = p2pclient.Client(
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._daemon_listen_port}'),
+            Multiaddr(f'/ip4/127.0.0.1/tcp/{self._client_listen_port}'))
+
+    async def _identify_client(self, delay):
+        await asyncio.sleep(delay)
+        encoded = await self._client.identify()
+        self.id = encoded[0].to_base58()
+
+    def _assign_daemon_ports(self, host_port=None, daemon_listen_port=None):
+        self._host_port, self._daemon_listen_port = host_port, daemon_listen_port
+        if host_port is None:
+            self._host_port = find_open_port()
+        if daemon_listen_port is None:
+            self._daemon_listen_port = find_open_port()
```

Comment:
Do you need this if you're checking for equality later and `host_port` can't be None at that point?

Reply:
I rewrote this to assigning later